// @ngx-lite/progress/esm2015/ngx-progress.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { NgxProgressBarComponent } from './ngx-progress-bar/ngx-progress-bar.component';
import { NgxProgressCircleComponent } from './ngx-progress-circle/ngx-progress-circle.component';
import * as i0 from "@angular/core";
export class NgxProgressModule {
}
NgxProgressModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0, type: NgxProgressModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgxProgressModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0, type: NgxProgressModule, declarations: [NgxProgressBarComponent, NgxProgressCircleComponent], imports: [CommonModule], exports: [NgxProgressBarComponent, NgxProgressCircleComponent] });
NgxProgressModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0, type: NgxProgressModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0, type: NgxProgressModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [NgxProgressBarComponent, NgxProgressCircleComponent],
                    exports: [NgxProgressBarComponent, NgxProgressCircleComponent],
                }]
        }] });
