// ack-angular-mobiscroll/__ivy_ngcc__/dist/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var common_1 = require("@angular/common");
var mobiscroll_components_1 = require("./mobiscroll.components");
var MbscProvider_1 = require("./MbscProvider");
__export(require("./mobiscroll.components"));
var ɵngcc0 = require('@angular/core');
var ɵngcc1 = require('./mobiscroll.components');
var ɵngcc2 = require('@angular/common');
var MbscModule = (function () {
    function MbscModule() {
    }
MbscModule.ɵfac = function MbscModule_Factory(t) { return new (t || MbscModule)(); };
MbscModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: MbscModule });
MbscModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [MbscProvider_1.MbscProvider], imports: [[
            common_1.CommonModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(MbscModule, [{
        type: core_1.NgModule,
        args: [{
                imports: [
                    common_1.CommonModule
                ],
                providers: [MbscProvider_1.MbscProvider],
                declarations: mobiscroll_components_1.declarations,
                exports: mobiscroll_components_1.declarations
            }]
    }], function () { return []; }, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(MbscModule, { declarations: [ɵngcc1.MobiscrollCalendar, ɵngcc1.MobiscrollDate, ɵngcc1.MobiscrollDateTime, ɵngcc1.MobiscrollTime], imports: [ɵngcc2.CommonModule], exports: [ɵngcc1.MobiscrollCalendar, ɵngcc1.MobiscrollDate, ɵngcc1.MobiscrollDateTime, ɵngcc1.MobiscrollTime] }); })();
    return MbscModule;
}());
/** @nocollapse */
MbscModule.ctorParameters = function () { return []; };
exports.MbscModule = MbscModule;
var MbscProvider_2 = require("./MbscProvider");
exports.MbscProvider = MbscProvider_2.MbscProvider;
var mobiscroll_components_2 = require("./mobiscroll.components");
exports.components = mobiscroll_components_2.declarations;
